pytest_redis.py

```diff
@@ -0,0 +1,63 @@
+"""pytest-redis queue plugin implementation."""
+
+import redis
+
+__version__ = "0.2.4"
+
+
+def pytest_addoption(parser):
+    """Add command line options to py.test command."""
+    parser.addoption('--redis-host', metavar='redis_host',
+                     type=str, help='The host of the redis instance.',
+                     required=True)
+    parser.addoption('--redis-port', metavar='redis_port',
+                     type=str, help='The port of the redis instance.',
+                     required=True)
+    parser.addoption('--redis-pop-type', metavar='redis_pop_type',
+                     type=str,
+                     help=('The key of the redis list containing '
+                           'the test paths to execute.'),
+                     default="RPOP")
+    parser.addoption('--redis-list-key', metavar='redis_list_key',
+                     type=str,
+                     help=('The key of the redis list containing '
+                           'the test paths to execute.'),
+                     required=True)
+
+
+def retrieve_test_from_redis(redis_connection, list_key, command):
+    """Remove and return a test path from the redis queue."""
+    val = None
+    if command.lower() == "lpop":
+        val = redis_connection.lpop(list_key)
+    elif command.lower() == "rpop":
+        val = redis_connection.rpop(list_key)
+    return val
+
+
+def pytest_cmdline_preparse(config, args):
+    """
+    Called before cmd line options are parsed.
+
+    We jump the gun and extract redis_host, redist_port
+    and redis_list_key from the cmdline args to start
+    pulling tests off the queue and add the tests to
+    the cmd line.
+    """
+    parse_args = config._parser.parse_known_args(args)
+
+    r_client = redis.StrictRedis(host=parse_args.redis_host,
+                                 port=parse_args.redis_port)
+
+    val = retrieve_test_from_redis(r_client,
+                                   parse_args.redis_list_key,
+                                   parse_args.redis_pop_type)
+    if val is None:
+        print "No items in redis queue '%s'" % parse_args.redis_list_key
+        print "Running all tests"
+
+    while val is not None:
+        args.append(val)
+        val = retrieve_test_from_redis(r_client,
+                                       parse_args.redis_list_key,
+                                       parse_args.redis_pop_type)
```

Comment:
This looks like it will:
  1. Start up and pop off a set of tests to run
  2. Add those tests to the args
  3. Run all tests selected

That works, but, this will probably be imbalanced in a parallelized build.  The first few test runners to start up will obtain most of the tests, leaving none for the others.  To more evenly distribute work, it's desirable to pop off a test, run it, and then fetch another (if any are left) only after it has completed running the first test.  This is desirable because, if tests are ordered in the queue from longest-duration to shortest, e.g. tests with duration (d) d=5, d=4, d=3, d=2, d=1 then with two test runners we would encourage a scenario where they would execute as follows:
  1. t=0 Runner 1 pops off d=5, Runner 2 pops off d=4
  1. t=4 Runner 2 pops off d=3
  1. t=5 Runner 1 pops off d=2
  1. t=7 Runner 1 pops off d=1, Runner 2 exits
  1. t=8 Runner 1 exits